### PR TITLE
Encrypt/Decrypt-ion Functionality Changes Using OpenSSL3.0 APIs 

### PR DESCRIPTION
**Pull request by vpatel9305**
server_transport encrypt and decrypt functionality changes using openSSL3.0 APIs
Added code under #If OPENSSL_VERSION_NUMBER condition, to support code compilation in ubi8 and ubi9 both images

### CODE REVIEW

File: server_transport/src/wstcp.c [hunk 1]
```diff
@@ -3321,7 +3321,7 @@ static char * http_escapepath(char * out, const char * in) {
     return ret;
 }
 
-const uint8_t keybase [] = {
+const unsigned char keybase [] = {
```
Comment (jonquark) [inline server_transport/src/wstcp.c:3324]: uint8_t is a better type, more in keeping with what we do elsewhere. I think this (and all the other uint changes)  should be reverted. What motivated the change?

File: server_transport/src/wstcp.c [hunk 2]
```diff
@@ -3369,6 +3370,49 @@ xUNUSED static const char * zz_encrypt(const char * data, char * buf, int len) {
...
+
+#else
+
+    EVP_CIPHER_CTX *ctx;
+    int len;
+    int ciphertext_len;
+
+    /* Create and initialise the context */
+    if(!(ctx = EVP_CIPHER_CTX_new()))
+        TRACE(7, "EVP_CIPHER_CTX_new failed\n");
```
Comment (jonquark) [inline server_transport/src/wstcp.c:3382]: Trace level is set to 5 by default i.e. any trace messages with number 6 or higher aren't recorded anywhere). It's a rare case but if these openssl functions fail, no ret code is returned (which would be a bigger change) and bad/no data is written to the buffer - if anyone was to try and debug this having at least some messages in the trace would help.

At a minimum lower the trace level, but how big a change would it be to make the function fail if the evp functions fail and the caller do something sensible with the error? If there is nothing sensible the caller can do consider creating an FFDC (a bunch of diagnostic data) and whether to make that FFDC fatal (i.e. stop the server) or not. Example ffdc:
https://github.com/eclipse/amlen/blob/main/server_admin/src/authentication.c#L1082

**Review by jonquark — COMMENTED** [review]
Looks most good - some comments on individual changes.

File: server_transport/src/wstcp.c (hunk 1 above)
Comment (vpatel9305) [reply server_transport/src/wstcp.c:3324]: openSSL3.0 exposed APIs to symmetric encryption/decryption are accepting unsigned char form of key and iv

`int EVP_EncryptInit_ex(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *type,
                       ENGINE *impl, const unsigned char *key, const unsigned char *iv);`

I found, having key and iv defining in the same data type would be helpful here and discard unnecessary type conversation.

Please let me know, if we still requires key and iv in uint type.

File: server_transport/src/wstcp.c (hunk 2 above)
Comment (vpatel9305) [reply server_transport/src/wstcp.c:3382]: I see, lowering trace level below or equal to 5 is useful in that case.

FFDC won't require here as we do not want to do any sensible with the error as if encrypt/decrypt fails then caller function must handle that failure, not the EV failure requires to do anything else than logging it.

As a part of this comment, I will change trace level to 5. Please suggest your thought.

Comment (jonquark) [reply server_transport/src/wstcp.c:3382]: At the moment - the caller doesn't know - there's no return code to say the encryption failed.

Comment (vpatel9305) [reply server_transport/src/wstcp.c:3382]: I see none of the caller working on failure of encrypt or decrypt function.
Do you think, aborting server on any EVP API call failure is good idea here ?

File: server_transport/src/wstcp.c (hunk 1 above)
Comment (vpatel9305) [reply server_transport/src/wstcp.c:3324]: As discussed in call, used unsigned char in old and new implementation (ubi8 as well as ubi9)

File: server_transport/src/wstcp.c (hunk 2 above)
Comment (vpatel9305) [reply server_transport/src/wstcp.c]: returning NULL on each EVP function failure.

**Review by IanBoden — APPROVED** [review]

**Pull request merged**